### ISSUE TITLE
How to plot my predicted candles in FreqUI?

### ISSUE THREAD

**Issue by LinXueyuanStdio** [freqtrade/freqtrade#8568]
## Describe your environment

  * Operating system: ubuntu
  * Python Version: 3.8 (`python -V`)
  * CCXT version: 3.0.36 (`pip freeze | grep ccxt`)
  * Freqtrade Version: 2023.4 (`freqtrade -V` or `docker compose run --rm freqtrade -V` for Freqtrade running in docker)
  
## Your question

*Ask the question you have not been able to find an answer in the [Documentation](https://www.freqtrade.io/en/latest/)*

I try to predict the future and run a timeseries prediction model to append the prediction to the dataframe. However, the plot can't show in the web. There are not any docs about the plotting config in FreqUI. I would like to plot my candles (up is orange, down is blue) and show the indicators in my dataframe. Whtat can I do?

By the way, the freqai is extremely hard to use. The strategy interface is not pure. Everything from ai to strategy has to be saved in dataframe, which is too stupid. Too much things are delegated to dk. I have to override start(dataframe, metadata, strategy) and do all the things myself. The freqai doesn't help.

**Comment by LinXueyuanStdio** [comment]
I want to plot the candles. But now I can only plot line.
![image](https://user-images.githubusercontent.com/23211526/235365001-ccbcc5b6-ea07-4699-87aa-11b2e6c80dbf.png)

**Comment by robcaulk** [comment]
Hello,

I am sorry that you find FreqAI to be "too stupid." We tried our best to share an interesting software with you, but it seems your level of intellect is just too far beyond ours. 

WRT your question:

> How to plot my predicted candles in FreqUI?

FreqAI creates a column in your dataframe with the predicted values. So plotting the prediction column is the same as plotting any other indicator you might create with TAlib. You could say that FreqAI is a supercharged indicator for the strategy. It means that:

```python
df["rsi"] = ta.RSI(df, 10)
```
creates a column of `rsi` that you can plot.

```python
df = self.freqai.start(df, metadata, self)
```

Gives you back a dataframe containing your prediction. If you are using the example strategy, that column is `&s-close`. Which you can plot the same way as you plot `rsi`.

So to plot the prediction it is the same whether you plot `rsi` or `&s-close`, go to your FreqUI gear in the top right, then add a plot in the GUI, add an indicator to the plot, select your prediction column (or select any other typical indicator). 

But, *let's be real*, I bet a man of your intellect is too bored to be clicking around in GUIs.  You probably prefer to copy paste the included plot config from the example strategy:

```python
    plot_config = {
        "main_plot": {},
        "subplots": {
            "&-s_close": {"prediction": {"color": "blue"}},
            "do_predict": {
                "do_predict": {"color": "brown"},
            },
        },
    }
```

Cheers,

Rob

**Comment by LinXueyuanStdio** [comment]
Hello, robcualk,

Thanks for responding. You are so kind and careful. I understand your words. But I know how to 'plot the prediction'. In fact, I successfully show the 'line' type of plot in FreqUI. My problem is how to show 'candle' type of plot, because my timeseries prediction model also predicts 'high' and 'low' prices, not only 'close' price. I do not find the setting of changing to 'candle' type in the GUI.

Then, I clarify that it is that 'Everything from ai to strategy has to be saved in dataframe' instead of freqAI that upsets me. In other words, the design of IFreqaiModel and IStrategy interfaces. These two interfaces are not independent at all.

I write my timeseries prediction model which handles its own normalization method specifically.
But It always fails when the code runs to predict() function. Finally I found that the ai interface supposes the subclass should not consider normalization. In fact, the IFreqaiModel performs normalization and denormalization by assigning "{label}-mean" and "{label}-std" columns to the dataframe. When I return my dataframe, processed by neural network model, in predict() function, it throws the error that "{label}-mean" and "{label}-std" columns not found??? I have to override them all.

About feature engineering, I doubt the need to separate xxx_expand_all, xx_expand_basic, xxx_standard. The three functions actually do the same thing. It may indicate that the strategy interface is over-designed. In my opinion, placing feature engineering in populate_indicator() is enough. Or user can design their feature engineering in ai model class, instead of  strategy class. The strategy just throws the raw dataframe OHLCV to ai and receives the processed dataframe+custom outputs from ai. The freqAI could be another independent project, while the freqtrade project provides as many APIs as it can for ai strategy. 

About dk.do_predict, I don't know how it works even after I read the documents. The IFreqaiModel merges my prediction dataframe by considering do_predict column in the dataframe. And the look back window crashes to 1 after one merge. At the same time, the new look back window also destroys my prediction model. I predict 96 time steps in the future, but it seems to only accept one time step prediction.

Sincerely,

Xueyuan

**Comment by robcaulk** [comment]
Hello,

I guess you missed the docs.

`label_mean` and `label_std` have nothing to do with normalization (why would it? normalization is just scaling between -1 and 1....are you confusing standardization and normalization or something?). `label_mean` and `label_std` are the mean and std values of your labels from the most recent training data. This information is extremely useful for deciding the rarity of the prediction relative to the training parameter space. This can enable dynamic entry/exits in the strategy. We wrote a guide on this for you: https://www.freqtrade.io/en/stable/freqai-configuration/#creating-a-dynamic-target-threshold. I guess you missed that. 

The three feature engineering functions are not the same. Please have a look at our documentation where we have explained in detail how these functions help users rapidly build 10k feature sets with single lines of code https://www.freqtrade.io/en/stable/freqai-feature-engineering/. Seems you missed that.

You seem mistaken about the role of Freqtrade and FreqAI. Freqtrade is not a library. FreqAI is not a library. FT combines with FreqAI to make a user-friendly application that both non-ML experts and power-users alike can exert adaptive modeling control on chaotic markets. A sandbox to start playing with open-source ML on adaptive modeling challenges. FreqAI can be controlled by config only, or by full manipulation of the IFreqaiModel. FreqAI is flexible, unlike your software engineering perspective.

`do_predict` is one of the most powerful and well-used tools that we have created. It enables powerful outlier detection and prediction confidence. The Dissimilarity Index is one of the most useful tools that FreqAI provides underneath `do_predict`. If you prefer not to use outlier detection, sure no problem, turn it off in your config. 

I dont know what it means "destroys prediction model." FreqAI trains at the exact frequency that you tell it to with `live_retrain_hours`. If you dont want adaptive modeling, then turn it off...?

I am closing this issue as your plotting issue seems unrelated to your real complaints about the software. You are welcome to submit PRs to our opensource repository to improve/change any part of Freqtrade/FreqAI that you wish (I wont hold my breath, you seem more interested in complaining than coding ;) ). 

Cheers,

Robert

**Comment by xmatthias** [comment]
>  In fact, I successfully show the 'line' type of plot in FreqUI. My problem is how to show 'candle' type of plot, because my timeseries prediction model also predicts 'high' and 'low' prices, not only 'close' price. I do not find the setting of changing to 'candle' type in the GUI.

FreqUI does plot the default candles (ohlcv) automatically.
It doesn't support plotting (additional) candles using custom columns, and we currently don't plan on adding support for this.
You can use "line",  "bar" or "scatter" plot types for your plots.
